### ISSUE TITLE
model.fit with dataset generator results in deadlock/hang 

### ISSUE THREAD

**Issue by talipini** [tensorflow/tensorflow#50637]
TF: 2.5.0
Python: 3.7.10

 Model.fit hangs during training. It appears dataset generator seems go into a deadlock while reading tfrecords. A simplified naive model is in the github gist below which reproduces the issue. Directly iterating on the generator seems to work fine but feeding it to model.fit seems to deadlock. Appreciate any color. 

https://gist.github.com/talipini/d82d1fa2f5d6b46f3222f2367570543a#file-tf-model-fit-hangs-with-dataset-generator-ipynb

**Comment by abattery** [comment]
Mutable variable support is now available in the nightly when converting using from_saved_model.
You need to set this flag to True for now.

converter.experimental_enable_resource_variables = True

**Comment by talipini** [comment]
Anyone have any workarounds or insights on this?

**Comment by UsharaniPagadala** [comment]
@talipini 

We see that the issue [template](https://github.com/tensorflow/tensorflow/issues/new?assignees=&labels=type%3Abug&template=00-bug-issue.md) has not been filled, could you please do so as it helps us analyse the issue [tf version, steps followed before you ran into this error or stand alone code to reproduce the issue faced]
Thanks

**Comment by talipini** [comment]
I have added github gist link which has the reproducible code to the original post but I will add the details here. 


**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Colab jupyter notebok.
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): Default version on Colab.
- TensorFlow version (use command below): 2.5.0
- Python version: 3.7.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: No GPU
- GPU model and memory: No GPU 

Dataset generator hangs when reading tfrecords files. This seems to happen after the first set of tfrecord files have been iterated through and starts reading the next block. This seems to happen consistently when using Colab default setup. 

**Describe the expected behavior**
Not hang.  

**Standalone code to reproduce the issue**
Link to the colab is here - https://colab.research.google.com/gist/talipini/b164e108501f831aff50827ac454d562/tf-model-fit-hangs-with-dataset-generator.ipynb

`
import tensorflow as tf
import numpy as np
import pandas as pd
import glob
import os
from pathlib import Path
used_compression_type = 'GZIP'

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Activation, Dense
from tensorflow.keras.optimizers import SGD

#Wrapper
class TFRecordConverter(object):
    def __init__(self):
        pass
    def int64_feature(self, value):
        return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))

    def float_feature(self, value):
        return tf.train.Feature(float_list=tf.train.FloatList(value=value))

    def bytes_feature(self, value):
        return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))

#Write Sample tfrecords

batch_size = 5
tf_archive = "/tfrecords"
os.makedirs(tf_archive, exist_ok=True)
print(f"TF Records at : {tf_archive}")

def create_test_tfrecords():
    existing_tfrecords = glob.glob(os.path.join(tf_archive,"*"))
    if len(existing_tfrecords) < 1:
        for i in range(0,5):
            x_data = np.random.uniform(low=-1.0, high=5, size= (batch_size*5, 6))
            write_to_TFRecords(os.path.join(tf_archive,f"tffile-{i}.tfr"),x_data)

def create_TFRecordsFile(file):
    Path(os.path.dirname(file)).mkdir(parents=True, exist_ok=True)    
    compression = tf.io.TFRecordOptions(compression_type = used_compression_type)
    writer =  tf.io.TFRecordWriter(file, options=compression)
    return writer

def write_to_TFRecords(file, values):
    bucket_writer = create_TFRecordsFile(file)
    shape = np.array(values.shape, np.int32)
    example = tf.train.Example(
            features = tf.train.Features(
               feature = {
                    'data':TFRecordConverter().float_feature(values.ravel()), #Float list can only be 1D
                    'shape':TFRecordConverter().bytes_feature(shape.tobytes())
                    }
               ))
    
    print(f"Writing to {file} with size: {shape}")
    bucket_writer.write(example.SerializeToString())
    

def parse_function(serialized_example):       
        
    features = {
        'data': tf.io.FixedLenSequenceFeature([], tf.float32,allow_missing=True),
        'shape':tf.io.FixedLenFeature([], tf.string)
        }

    features = tf.io.parse_single_example(serialized=serialized_example, features=features)
    shape = tf.io.decode_raw(features['shape'], tf.int32 )
    dataset = features['data']
    dataset = tf.reshape(dataset, shape)


    return dataset

#Simple dataset generator
def data_gen(batch_size):
    tfrecords = glob.glob(os.path.join(tf_archive,"*"))
    print(f"Loading tfrtecords: {tfrecords}")
    dataset = tf.data.Dataset.from_tensor_slices(tfrecords) 
    dataset = dataset.interleave(lambda x: tf.data.Dataset.from_generator(gen_step, 
                                output_types=( tf.float32, tf.float32), args=(x,)))
        
    dataset = dataset.repeat().batch(batch_size)
    while True:
        for x, Y in dataset:
            yield x, Y

def gen_step(tf_file):
    trRecordDataset = tf.data.TFRecordDataset(tf_file, compression_type=used_compression_type)
    trRecordDataset = trRecordDataset.map(parse_function)
    print(f"Waiting for dataset in step - {tf_file}")
    for dataset in trRecordDataset: 
        Y = dataset[:,0]
        x = dataset[:,1:]
        count = 0
        while count < dataset.shape[0]:
            yield x[count],Y[count]
            count += 1
    print(f"Finsihed dataset in - {tf_file}")

if __name__ == '__main__':
  create_test_tfrecords()
  #trivial model - hangs in 
  model = Sequential()
  model.add(tf.keras.Input(shape=(batch_size,)))
  model.add(Dense(1, activation='linear'))        
  model.compile(loss='categorical_crossentropy', optimizer=SGD(), metrics = ['accuracy'])
  
  #model.fit will hang just as it starts to read the second set of tfrecords.
  model.fit(x = data_gen(batch_size=batch_size),
          steps_per_epoch=5,
          epochs=10)
  print(f"Training Finished.")
`

**Comment by talipini** [comment]
This could be related to [43768](https://github.com/tensorflow/tensorflow/issues/43768) but not sure.

**Comment by UsharaniPagadala** [comment]
@talipini 

Could you Please refer [this](https://stackoverflow.com/questions/47705684/tensorflow-tf-data-dataset-from-generator-does-not-work-with-strings-on-pyt), hope it helps.Thanks

**Comment by talipini** [comment]
Thank you. First one is referring to an issue with strings in the generator, which I am not using. The [second](https://github.com/tensorflow/tensorflow/issues/21277) one is referring to deadlock on session destructor but that is supposed to be fixed in 2018. I went through both chains but I am kinda lost on it.  But [43768](https://github.com/tensorflow/tensorflow/issues/43768) looks interesting and could be related.

**Comment by talipini** [comment]
Any additional insights on this? thanks

**Comment by UsharaniPagadala** [comment]
@rmothukuru 

I was able to replicate the issue reported here.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/2c404dad2b1abe24fab6f75ad5ab0df0/tf-model-fit-hangs-with-dataset-generator.ipynb).Thanks

**Comment by talipini** [comment]
Thank you. Appreciate any color on this.

**Comment by talipini** [comment]
@rmothukuru , @aaudiber - Any updates on this? Appreciate any insights. Thanks

**Comment by grofte** [comment]
This appears to have been fixed in version 2.7

**Comment by chunduriv** [comment]
@talipini,

Sorry for the late response. This issue is fixed now. 

```
#working now 
model = Sequential()
model.add(tf.keras.Input(shape=(batch_size,)))
model.add(Dense(1, activation='linear'))        
model.compile(loss='categorical_crossentropy', optimizer=SGD(), metrics = ['accuracy'])

 
model.fit(x = data_gen(batch_size=batch_size),
        steps_per_epoch=5,
        epochs=10)
print(f"Training Finished.")
```

Output:
```
Loading tfrtecords: ['/tfrecords/tffile-0.tfr', '/tfrecords/tffile-1.tfr', '/tfrecords/tffile-2.tfr', '/tfrecords/tffile-4.tfr', '/tfrecords/tffile-3.tfr']
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/impl/api.py:458: calling DatasetV2.from_generator (from tensorflow.python.data.ops.dataset_ops) with output_types is deprecated and will be removed in a future version.
Instructions for updating:
Use output_signature instead
Waiting for dataset in step - b'/tfrecords/tffile-0.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-1.tfr'
Epoch 1/10
5/5 [==============================] - 2s 26ms/step - loss: 2.5224e-07 - accuracy: 0.0000e+00
Epoch 2/10
5/5 [==============================] - 0s 28ms/step - loss: 2.1971e-07 - accuracy: 0.0000e+00
Epoch 3/10
Finsihed dataset in - b'/tfrecords/tffile-0.tfr'
Finsihed dataset in - b'/tfrecords/tffile-1.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-2.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-4.tfr'
5/5 [==============================] - 1s 18ms/step - loss: 3.0705e-07 - accuracy: 0.0000e+00
Epoch 4/10
5/5 [==============================] - 0s 20ms/step - loss: 2.3033e-07 - accuracy: 0.0000e+00
Epoch 5/10
Finsihed dataset in - b'/tfrecords/tffile-2.tfr'
Finsihed dataset in - b'/tfrecords/tffile-4.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-3.tfr'
5/5 [==============================] - 1s 26ms/step - loss: 2.2736e-07 - accuracy: 0.0000e+00
Epoch 6/10
Finsihed dataset in - b'/tfrecords/tffile-3.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-0.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-1.tfr'
5/5 [==============================] - 1s 26ms/step - loss: 2.5224e-07 - accuracy: 0.0000e+00
Epoch 7/10
5/5 [==============================] - 0s 17ms/step - loss: 2.1971e-07 - accuracy: 0.0000e+00
Epoch 8/10
Finsihed dataset in - b'/tfrecords/tffile-0.tfr'
Finsihed dataset in - b'/tfrecords/tffile-1.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-2.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-4.tfr'
5/5 [==============================] - 1s 28ms/step - loss: 3.0705e-07 - accuracy: 0.0000e+00
Epoch 9/10
5/5 [==============================] - ETA: 0s - loss: 2.3033e-07 - accuracy: 0.0000e+00Finsihed dataset in - b'/tfrecords/tffile-2.tfr'
5/5 [==============================] - 0s 30ms/step - loss: 2.3033e-07 - accuracy: 0.0000e+00
Finsihed dataset in - b'/tfrecords/tffile-4.tfr'
Epoch 10/10
Waiting for dataset in step - b'/tfrecords/tffile-3.tfr'
5/5 [==============================] - 0s 18ms/step - loss: 2.2736e-07 - accuracy: 0.0000e+00
Finsihed dataset in - b'/tfrecords/tffile-3.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-0.tfr'
Waiting for dataset in step - b'/tfrecords/tffile-1.tfr'
Training Finished.
```
Please find the [gist](https://colab.sandbox.google.com/gist/chunduriv/1b7ffd5fce0ee37333ff44d1138cd464/tf-model-fit-hangs-with-dataset-generator.ipynb) for reference. Thank you.

**Comment by OysterQAQ** [comment]
same problem
